CASSANDRA-17062 only cache refactoring

smiklosovic opened the pull request:
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

Mmuzaf commented on src/java/org/apache/cassandra/cache/UnweightedCacheSize.java:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * Specifies the maximum number of entries in this cache.
+     */
+    void setMaxEntries(int maxEntries);
```
This method is not used.

smiklosovic replied:
is not it used in `AuthCache#setMaxEntries`? It basically implements that method which is call elsewhere, like in `SetAuthCacheConfig`.

smiklosovic replied:
hah so "problem" here is that `AuthCache` implements `AuthCacheMBean` which has a method of the same signature and then it changes the size via `setMaxEntriesDelegate`. I am not going deeper than this. I think this is just fine ... for cases when a cache would not extend MBean with such method name, we would still need a way how to change max entries etc ...

Mmuzaf commented on src/java/org/apache/cassandra/metrics/AbstractCacheMetrics.java:
```diff
@@ -0,0 +1,113 @@
+import static org.apache.cassandra.metrics.CassandraMetricsRegistry.Metrics;
+
+/**
+ * Metrics for an unweighted cache like {@code AuthCache}.
```
The comment is not accurate - this class is used to hold both unweighted and weighted cache metrics.

Mmuzaf commented on src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * Metrics for an unweighted cache like {@code AuthCache}.
+ */
+public class UnweightedCacheMetrics extends AbstractCacheMetrics
```
It took me a while to figure out the difference between weighted and unweighted terms, so I assume the users will have to do the same exercise. At the moment,the `UnweightedCache` fully matches the `AuthCache`, so maybe we should classify these types of caches in this way, wdyt? This should be easier to understand for a user.

smiklosovic replied:
@Mmuzaf are AuthCaches the only caches ever which will be unweighted? I think the general distinction between weighted and unweighted makes sense. Maybe we could extra document that? Othewise I think it is just ok as it is ...

Mmuzaf replied:
In the future, however, weighted and unweighted may not be sufficient for us as well, so I wouldn't want to introduce a new concept at this point, as the `AuthCache` fully covers all our needs. It also sounds clearer without adding pages of documentation, and we can also change it in the future if a new type arises and we will decide to classify it... :-)

smiklosovic replied:
I have a hard time to figure out what other metrics we would have so we could not categorize them into weighted or unweighted ones.

Mmuzaf replied:
AuthCache metrics are valuable in their own right, though, so I'm not sure we need an additional 'unweighted' category for them. The simpler it is, the better it will be. 
We can move this question to the issue for visibility. 
@beobal I think it was your original idea to organise classes in this way. wdyt ?

beobal replied:
I tend to agree with @smiklosovic that the general distinction between weighted and unweighted makes sense.

> In the future, however, weighted and unweighted may not be sufficient for us as well, so I wouldn't want to introduce a new concept at this point, as the AuthCache fully covers all our needs.

I don't think any new concept is being introducing here, unweighted caches exist in C* you are just exposing them via the new interface. 

I also don't agree that this requires pages of additional documentation; _some_ docs will be required to direct users to the appropriate metrics, so it seems to me a choice between 

`"For metrics related to auth caches, see the AuthCache metrics 'nameX, nameY, nameZ...'"`
vs
`"For metrics related auth caches, see the UnweightedCache metrics 'nameX, nameY, nameZ...'"`

I don't believe that docs for the metrics themselves don't need to go into great detail about the nature of weighted vs unweighted caches.

That's my $0.02, but I don't feel mega-strongly about it so wouldn't object whichever way you decide.

Mmuzaf commented on src/java/org/apache/cassandra/cache/ICache.java:
```diff
@@ -23,6 +23,8 @@
  * and does not require put or remove to return values, which lets SerializingCache
  * be more efficient by avoiding deserialize except on get.
+ *
+ * This class represents a weighted cache.
```
There is also a cache called `ChunkCache' which also has a weighted type, but it doesn't extend the `ICache' interface. So, we either have to fix javadocs for both of these cases or for none of them. The latter case sounds easier to me as the `CacheSize` already has a description.

Mmuzaf commented on src/java/org/apache/cassandra/cache/ChunkCache.java:
```diff
@@ -172,7 +171,7 @@ public void onRemoval(Key key, Buffer buffer, RemovalCause cause)
     }
 
-    public void close()
+    public void clear()
```
This change is not related to the issue, right?

smiklosovic replied:
I cant find any usage of that method. You are technically right, yes. But looking at it, why not to rename it to `clear`?
 The impl invalidates the cache which is kind of a clearing ... I think "clear" makes more sense and reflects what that method does than "close". That cache does not implement `(Auto)Closeable` or anything like that, there is nothing to _close_. I think that was the original idea behind this change. Keep in mind I just put it all together after Sam did his improvements.

Mmuzaf commented on src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java:
```diff
@@ -0,0 +1,53 @@
+    {
+        super(new DefaultNameFactory("UnweightedCache", type), type);
+
+        maxEntries = Metrics.register(factory.createMetricName("Capacity"), cache::maxEntries);
```
Should we still keep it synced with the `CacheMetrics` names? I think we can use the "maxEntries" instead of the "Capacity" to avoid confusion.

smiklosovic replied:
Is this change only internally visible? If it is exposed to clients I think we need to keep it like it is to not break anything.

Mmuzaf replied:
This will be exposed to an end user, however, as we don't have such metrics in the public API at this moment, we can expose it with the "right" name. Even though auth caches and weighted caches have different MBeans, vtables there is no need to keep the attribute names in sync between them, so I guess we can change it.

smiklosovic replied:
@Mmuzaf I would rather stay consistent and rename `maxEntries` gauge to `capacity` to be on par with `CacheMetrics` which were there first which have it `capacity` too. There is no reason to have this naming discrepancy among different types of cache metrics and term "capacity" is just perfectly fine.

smiklosovic replied:
you know what ... you are right. `Capacity` here should be `MaxEntries`. Because `Capacity` and `MaxEntries` are not the same thing.

Mmuzaf commented on src/java/org/apache/cassandra/tools/NodeProbe.java:
```diff
@@ -1748,22 +1749,22 @@ public Object getCacheMetric(String cacheType, String metricName)
                     return JMX.newMBeanProxy(mbeanServerConn,
-                            new ObjectName("org.apache.cassandra.metrics:type=Cache,scope=" + cacheType + ",name=MissLatency"),
-                            CassandraMetricsRegistry.JmxTimerMBean.class).getDurationUnit();
+                                             new ObjectName("org.apache.cassandra.metrics:type=" + type + ",scope=" + scope + ",name=MissLatency"),
```
Should we change `name=MissLatency` to `name=MissLatencyUnit` or am I missing something?

Mmuzaf commented on src/java/org/apache/cassandra/tools/NodeProbe.java:
```diff
@@ -1734,10 +1734,11 @@ public long getReadRepairRepairedBackground()
      * @param metricName Capacity, Entries, HitRate, Size, Requests or Hits.
      */
-    public Object getCacheMetric(String cacheType, String metricName)
+    public Object getCacheMetric(String type, String scope, String metricName)
```
Since the `getCacheMetric` method isn't currently used for unweighted caches, would it be better to roll back these changes and do it when the nodetool needs it? 

Another option is to create an overloaded `public Object getCacheMetric` method that takes "Cache" as the first argument, this will help us avoid changes that are not necessary right now. The weighted and unweighted caches are slightly different anyway, so `getCacheMetric` is not quite what we want (e.g. the `size` is different for weighted and unweighted).

smiklosovic replied:
I think this is leftover from times when I was trying to expose unweighted caches as well into nodetool's `Info` so it was not always called with "Cache" but I removed that functionality from `Info` because I just felt in my guts that this is not right thing to do anyway ... Metrics are going to be exposed via CQL and adding more metrics into nodetool just feels wrong. I can revert this whole change.

azotcsit committed:
Refactor structure of caching metrics

patch by Aleksei Zotov; reviewed by Stefan Miklosovic, Maxim Muzafarov for CASSANDRA-17062

Co-authored-by: Sam Tunnicliffe <samt@apache.org>
Co-authored-by: Stefan Miklosovic <smiklosovic@apache.org>

Mmuzaf commented on doc/modules/cassandra/pages/managing/operating/metrics.adoc:
```diff
@@ -528,17 +528,17 @@ ViewReplicasSuccess.
 |===
 
-== Cache Metrics
+== Weighted Cache Metrics
```
Should we also add a few words about 'weighted' and 'unweighted' terms? I think here is the comment with a well-described difference, we can use it:

https://issues.apache.org/jira/browse/CASSANDRA-17062?focusedCommentId=17443143&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17443143

smiklosovic replied:
done

Mmuzaf commented on test/unit/org/apache/cassandra/auth/AuthCacheTest.java:
```diff
@@ -322,6 +322,41 @@ public void testCacheLoaderIsNotCalledOnGetAllWhenCacheIsNotEmpty()
     }
 
+    @Test
+    public void testMetricsOnCacheEnabled()
```
Should we also check the numbers for the `entries` and `maxEntries` method calls? They do not seem to be covered by the tests.

smiklosovic replied:
done

Mmuzaf commented on test/unit/org/apache/cassandra/auth/AuthCacheTest.java:
```diff
@@ -322,6 +322,41 @@ public void testCacheLoaderIsNotCalledOnGetAllWhenCacheIsNotEmpty()
+    }
+
+    @Test
+    public void testMetricsOnCacheDisabled()
```
Should we add a test to check that new metrics are available via the JMX API? I haven't found a good candidate for a test class, but I think it's still worth checking that `getMaxEntries` and `setMaxEntries` match, and the `entries` match the number entries we set to the cache.

smiklosovic replied:
I dont think that involving _JMX_ is necessary. We basically just need to invoke same MBean method programmatically. I have added "testSettingMaxEntries" test for that.

Mmuzaf commented on src/java/org/apache/cassandra/auth/AuthCache.java:
```diff
@@ -225,6 +235,7 @@ public V get(K k)
         if (cache == null)
             return loadFunction.apply(k);
 
+        metrics.requests.mark();
```
Aren't we missing the `getAll` path for tracking requests here? The `ChunkCache` itself encapsulates this logic in the `StatsCounter`.  It's worth adding such a test or a verification the metrics value for existing tests in `AuthCacheTest`.

smiklosovic replied:
Hmm ... good catch.

Mmuzaf commented on src/java/org/apache/cassandra/cache/UnweightedCacheSize.java:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * Returns the maximum number of entries in this cache.
+     */
+    int maxEntries();
```
Should this be renamed to `getMaxEntries`?

smiklosovic replied:
The naming of these methods basically copies these in `CacheSize`. I would keep it like it is. In `CacheSize` there is `capacity` / `getCapacity` etc ... We would need to rework everything, `int size()`, `long weightedSize()` to prefix with `get` etc ...

Mmuzaf commented on src/java/org/apache/cassandra/auth/AuthCache.java:
```diff
@@ -52,12 +58,13 @@
 
     public static final String MBEAN_NAME_BASE = "org.apache.cassandra.auth:type=";
 
+    @SuppressWarnings("rawtypes")
```
Maybe we should use `ScheduledFuture<?>` instead?

smiklosovic replied:
huh how did this happen ... I don't remember putting it there, sure, can be like suggested.

Mmuzaf reviewed (approved):
LGTM, +1
